### PR TITLE
AK, LibCompress: Add randomized tests

### PR DESCRIPTION
**Pull request by Janiczek**
Let me know if you want some of these split into separate PRs!
I have more tests to write, I'm going through the current test suites alphabetically. Probably to come in separate PRs.
What I have right now:

## Roundtrip properties

> dec(enc(x)) == x

- LibCompress/Gzip - `decompress_all`, `compress_all` (fuzzer)
- AK/Base64 - `decode_base64`, `encode_base64` (fuzzer)
- AK/BitStream - `read_bits`, `write_bits` (randomized test, ⚠️ on a single call though, BitStream behaves weird when multiple reads and writes happen)

## Oracle tests

> my_impl(x) == definitely_correct_impl(x)

- CharacterTypes
  - `isascii` == `is_ascii` in the UNICODE char range
  - `tolower` == `to_ascii_lowercase` in the UNICODE char range
  - `toupper` == `to_ascii_uppercase` in the UNICODE char range

## Safety/Liveness

> Safety = "doesn't do the wrong thing"
> Liveness = "does the right thing"

- BinaryHeap
  - `pop_min()` returns values in ASC sorted order
  - `pop_min()` returns the same values as `peek_min()`
- BinarySearch
  - finds number that is present in a vector
  - doesn't find number that is not present in a vector
- Bitmap
  - after `set`, `get` finds the value
  - after `set_range`, `get` finds the values and `count_in_range` reports the correct count
  - after `fill`, the whole bitmap is set with the given value
  - `find_one_anywhere` finds a value after we set a value
  - `find_first` finds a value after we set a value

## Other

- AllOf
  - an always true predicate makes `allOf` true for any input (even empty)
  - an always false predicate makes `allOf` false for any non-empty input
  - also unit tests for vacuous truth
- AnyOf
  - an always true predicate makes `anyOf` true for any non-empty input
  - an always false predicate makes `anyOf` false for any input (even empty)
  - also unit tests for vacuous truth
- BuiltinWrappers
  - `count_leading_zeroes` doesn't change if set bits below MSB
  - `count_required_bits(n) == log2(n) + 1`
  - `bit_scan_forward(n) == count_trailing_zeroes(n) + 1` (except for 0)

### CODE REVIEW

File: Userland/Libraries/LibTest/Randomized/Generator.h [hunk 1]
```diff
@@ -49,7 +49,9 @@ inline u32 unsigned_int(u32 max)
         return 0;
 
     u32 random = Test::randomness_source().draw_value(max, [&]() {
-        return AK::get_random_uniform(max + 1);
+        // `clamp` to guard against integer overflow and calling get_random_uniform(0).
+        u32 exclusive_bound = AK::clamp(max + 1, max, NumericLimits<u32>::max());
```
Comment (Janiczek) [inline Userland/Libraries/LibTest/Randomized/Generator.h:53]: @ADKaster I think this makes sense; might have misunderstood you on Discord though. This IMHO implements the "attempt to increment, but if you end up with 0, keep the original value" behaviour that I originally had `AK::max` do.

**Comment by Janiczek** [comment]
After discussion with @timschumi I'll change the roundtrip tests into fuzzers, since AFL can do this much better and we're operating on raw bytes instead of structured data.

File: Tests/AK/TestCharacterTypes.cpp [hunk 2]
```diff
@@ -34,7 +36,31 @@ void compare_value_output_over(u32 range, auto& old_function, auto& new_function
...
+void randomized_compare_value_output_over(u32 range, auto& old_function, auto& new_function)
+{
+    // NOTE: randomized tests also run multiple times (100 by default). This means we'll try 10k random numbers times each time the test suite is run.
+    for (u32 n = 0; n < 100; ++n) {
+        i64 result1 = false;
+        i64 result2 = false;
+        GEN(i, Gen::unsigned_int(range - 1));
+        EXPECT_EQ(result1 = old_function(i), result2 = new_function(i));
+        if (result1 != result2)
+            dbgln("Function input value was {}.", i);
```
Comment (timschumi) [inline Tests/AK/TestCharacterTypes.cpp:61]: We have a `FAIL` macro that one can pass a completely custom string into. This means something like the following should work:

```suggestion
        if ((result1 = old_function(i)) != (result2 = new_function(i))
            FAIL(String::formatted("New result {} does not match old result {} for input {}", result2, result1, i));
```

File: Tests/AK/TestCharacterTypes.cpp [hunk 3]
```diff
@@ -34,7 +36,31 @@ void compare_value_output_over(u32 range, auto& old_function, auto& new_function
...
+void randomized_compare_bool_output_over(u32 range, auto& old_function, auto& new_function)
+{
+    // NOTE: randomized tests also run multiple times (100 by default). This means we'll try 10k random numbers times each time the test suite is run.
+    for (u32 n = 0; n < 100; ++n) {
+        bool result1 = false;
+        bool result2 = false;
+        GEN(i, Gen::unsigned_int(range - 1));
+        EXPECT_EQ(result1 = (old_function(i) > 0), result2 = (new_function(i) > 0));
+        if (result1 != result2)
+            dbgln("Function input value was {}.", i);
```
Comment (timschumi) [inline Tests/AK/TestCharacterTypes.cpp:48]: Same here.

File: Tests/AK/TestCharacterTypes.cpp [hunk 4]
```diff
@@ -34,7 +36,31 @@ void compare_value_output_over(u32 range, auto& old_function, auto& new_function
...
+    }
+}
+
+void randomized_compare_value_output_over(u32 range, auto& old_function, auto& new_function)
+{
+    // NOTE: randomized tests also run multiple times (100 by default). This means we'll try 10k random numbers times each time the test suite is run.
+    for (u32 n = 0; n < 100; ++n) {
+        i64 result1 = false;
+        i64 result2 = false;
```
Comment (timschumi) [inline Tests/AK/TestCharacterTypes.cpp:57]: I guess this is a copy-paste mistake, since an `i64` shouldn't be initialized with a boolean value (at least logically)?

File: Tests/AK/TestBuiltinWrappers.cpp [hunk 5]
```diff
@@ -65,3 +67,39 @@ TEST_CASE(wrapped_count_required_bits)
...
+RANDOMIZED_TEST_CASE(count_required_bits)
+{
+    // count_required_bits(n) == log2(n) + 1
+
+    GEN(n, Gen::unsigned_int());
+    ASSUME(n > 0);
+    // log2(0) is -infinity, we don't care about that
+
+    size_t expected = max(1, AK::log2(static_cast<double>(n)) + 1);
+    // max due to edge case for n=0
```
Comment (timschumi) [inline Tests/AK/TestBuiltinWrappers.cpp:95]: Comments go above the code that they comment.

File: Tests/AK/TestBuiltinWrappers.cpp [hunk 6]
```diff
@@ -65,3 +67,39 @@ TEST_CASE(wrapped_count_required_bits)
...
+{
+    //    count_leading_zeroes(0b000...0001000...000)
+    // == count_leading_zeroes(0b000...0001___...___) (where _ is 0 or 1)
+    GEN(e, Gen::unsigned_int(0, 63));
+    auto power_of_two = 1ULL << e;
+    // 2^e
+
+    GEN(below, Gen::unsigned_int(0, power_of_two - 1));
+    auto n = power_of_two + below;
+    // 2^e + random bits below the first 1
```
Comment (timschumi) [inline Tests/AK/TestBuiltinWrappers.cpp:81]: Same here.

File: Meta/Lagom/Fuzzers/FuzzGzipRoundtrip.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCompress/Gzip.h>
+#include <stdio.h>
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
```
Comment (timschumi) [inline Meta/Lagom/Fuzzers/FuzzGzipRoundtrip.cpp:10]: Please rebase onto the latest `master`, all the fuzzers now have an additional call to `AK::set_debug_enabled(false)` to avoid printing debug output that slows down performance.

File: Meta/Lagom/Fuzzers/FuzzBase64Roundtrip.cpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Base64.h>
+#include <stdio.h>
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
```
Comment (timschumi) [inline Meta/Lagom/Fuzzers/FuzzBase64Roundtrip.cpp:10]: Same here regarding `set_debug_enabled`.

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Tests/LibCompress/TestGzip.cpp [hunk 9]
```diff
@@ -18,7 +18,7 @@ TEST_CASE(gzip_decompress_simple)
         0x00, 0xc2, 0x1d, 0x22, 0x15, 0x0f, 0x00, 0x00, 0x00
     };
 
-    const u8 uncompressed[] = "word1 abc word2";
+    u8 const uncompressed[] = "word1 abc word2";
```
[the remark this answers is not in the document]
Comment (Janiczek) [reply Tests/LibCompress/TestGzip.cpp:21]: Yeah they are leftovers from when I clang-formatted the file (due to writing the roundtrip test there initially). Do you want the change moved to its own commit, or removed altogether?

File: Tests/AK/TestCharacterTypes.cpp (hunk 4 above)
Comment (Janiczek) [reply Tests/AK/TestCharacterTypes.cpp:57]: You're right, I didn't spot this difference between the `compare_value_output` and `compare_bool_output`. Will fix

File: Tests/AK/TestBitmap.cpp [hunk 10]
```diff
@@ -286,6 +331,25 @@ TEST_CASE(find_one_anywhere_edge_case)
...
+    GEN(i, Gen::unsigned_int(size - 1));
+
+    auto bitmap = MUST(Bitmap::create(size, init));
+    bitmap.set(i, new_value);
+
+    Optional<size_t> result = new_value
+        ? bitmap.find_one_anywhere_set(hint)
+        : bitmap.find_one_anywhere_unset(hint);
+
+    EXPECT(result.has_value());
```
[the remark this answers is not in the document]
Comment (Janiczek) [reply Tests/AK/TestBitmap.cpp:350]: We have, roughly, three choices:

1. keep all the parameters completely random and don't check the specific index that's found and returned (since if you're setting `true` in an already fully `true` bitmap, the index that will be returned is 0, not the `i` at which you have set)
2. hardcode some of the parameters (ie. if new_value = !init) and be able to check the index since it will always be the generated `i`
3. like 1, randomize everything, but be a bit smart about what index you're expecting: if `new_value == init` then expected index == 0, else expected index == `i`.

Do you want (3)?

[the remark this answers is not in the document]
Comment (timschumi) [reply Tests/AK/TestBitmap.cpp:350]: (3) would be great, yes.

File: Userland/Libraries/LibTest/Randomized/Generator.h (hunk 1 above)
Comment (ADKaster) [reply Userland/Libraries/LibTest/Randomized/Generator.h:53]: Hm. I guess this is all right. I kind of meant something more like `u32 bound = 1 + AK::clamp(max, 0, NumericLimits::max() - 1)` but :shrug:

File: Tests/AK/TestBuiltinWrappers.cpp [hunk 11]
```diff
@@ -65,3 +67,40 @@ TEST_CASE(wrapped_count_required_bits)
...
+RANDOMIZED_TEST_CASE(count_required_bits)
+{
+    // count_required_bits(n) == log2(n) + 1
+
+    // log2(0) is -infinity, we don't care about that
+    GEN(n, Gen::unsigned_int());
+    ASSUME(n > 0);
+
+    // max due to edge case for n=0
+    size_t expected = max(1, AK::log2(static_cast<double>(n)) + 1);
```
Comment (timschumi) [inline Tests/AK/TestBuiltinWrappers.cpp:96]: Note that `AK::log2(0)` would still get evaluated, which is probably not great.

Shouldn't we just do `Gen::unsigned_int(1, NumericLimits<u32>::max());` here? That would presumably get rid of both the `ASSUME` line and the need to work around `log2(0)`.

File: Tests/AK/TestBuiltinWrappers.cpp [hunk 12]
```diff
@@ -65,3 +67,40 @@ TEST_CASE(wrapped_count_required_bits)
...
+
+    EXPECT_EQ(count_required_bits(n), expected);
+}
+
+RANDOMIZED_TEST_CASE(bit_scan_forward_count_trailing_zeroes)
+{
+    // behaviour for 0 differs so we skip it
```
Comment (timschumi) [inline Tests/AK/TestBuiltinWrappers.cpp:103]:
```suggestion
    // Behaviour for 0 differs, so we skip it.
```

Not sure about the comma, but comments should at the very least start with a capital letter and end with a period if they are meant to resemble a sentence (excluding variable names, or quotes from elsewhere).

**Review by timschumi — CHANGES_REQUESTED** [review]
Two remaining very small nitpicks.

File: Tests/AK/TestBuiltinWrappers.cpp (hunk 11 above)
Comment (Janiczek) [reply Tests/AK/TestBuiltinWrappers.cpp:96]: Oh, reading this again, we don't need the max at all anymore, thanks to the `ASSUME(n > 0)`. I think I can remove that!

As for how to generate,
```cpp
GEN(n, Gen::unsigned_int());
ASSUME(n > 0);
```
and
```cpp
GEN(n, Gen::unsigned_int(1, NumericLimits<u32>::max()));
```
are roughly equivalent in behaviour (ASSUME seems to me cleaner but more CPU-wasteful since it has to retry if it generates 0).

Comment (timschumi) [reply Tests/AK/TestBuiltinWrappers.cpp:96]: I'd prefer the non-`ASSUME` variant if possible, since that eliminates one more `ASSUME` that can fail randomly (although, arguably, the chances here are much smaller than with generating a boolean).

**Review by timschumi — APPROVED** [review]
